model-engine/model_engine_server/inference/batch_inference/vllm_batch.py

```diff
@@ -111,6 +117,161 @@ def delete_s3_chunks(request):
     print("Chunks deleted")
 
 
+def random_uuid() -> str:
+    return str(uuid.uuid4().hex)
+
+
+def get_vllm_engine(model, request):
+    from vllm import AsyncEngineArgs, AsyncLLMEngine
+
+    engine_args = AsyncEngineArgs(
+        model=model,
+        quantization=request.model_config.quantize,
+        tensor_parallel_size=request.model_config.num_shards,
+        seed=request.model_config.seed or 0,
+        disable_log_requests=True,
+    )
+
+    llm = AsyncLLMEngine.from_engine_args(engine_args)
+    return llm
+
+
+async def generate_with_tool(
+    llm,
+    tool_config: ToolConfig,
+    content: CreateBatchCompletionsRequestContent,
+    prompts,
+    tool: BaseTool,
+):
+    class IterativeGeneration:
+        def __init__(self, prompt, max_new_tokens):
+            self.generated_text = ""
+            self.num_prompt_tokens = 0
+            self.remaining_tokens = max_new_tokens
+            self.token_logits = []
+            self.tool_exception = None
+            self.prompt = prompt
+            self.completed = False
+
+        def __repr__(self) -> str:
+            return f"generated_text: {self.generated_text}, num_prompt_tokens: {self.num_prompt_tokens}, remaining_tokens: {self.remaining_tokens}, tool_exception: {self.tool_exception}, prompt: {self.prompt}, completed: {self.completed}"
```

Comment:
should we also log the `token_logits`?

Reply:
this is for debugging only (not printing this anyway). also there's too much data